File: LLMgeneratedCode/gemini-3.R
# Load necessary libraries
library(readr)
library(dplyr)
library(tidyr)
library(glmnet)
library(Metrics)
library(ggplot2)

# Define file paths
feature_file_training <- "df_training.csv"
feature_file_test <- "df_test.csv"
ano_file_training <- "ano_training.csv"
ano_file_test <- "ano_test.csv"

# Load feature data
df_training_feature <- read_csv(feature_file_training)
df_test_feature <- read_csv(feature_file_test)

# Load annotation data
ano_training <- read_csv(ano_file_training)
ano_test <- read_csv(ano_file_test)

# ------------------- Data Preprocessing -------------------

# --- Training Data ---
# Transpose feature data: Samples as rows, Features as columns
df_training_feature_t <- df_training_feature %>%
  pivot_longer(cols = -Feature_Name, names_to = "Sample_ID", values_to = "Methylation") %>%
  pivot_wider(names_from = Feature_Name, values_from = Methylation)

# Extract feature names (excluding Sample_ID column for feature data)
feature_names <- colnames(df_training_feature_t)[-1]

# Merge feature data with annotation data by Sample_ID
train_data <- ano_training %>%
  inner_join(df_training_feature_t, by = "Sample_ID")

# Extract Gestational Age (GA) as target variable
train_GA <- train_data$Gestational_Age

# Select feature columns (methylation features) for training
train_features <- train_data %>% select(all_of(feature_names))

# --- Test Data ---
# Transpose feature data: Samples as rows, Features as columns
df_test_feature_t <- df_test_feature %>%
  pivot_longer(cols = -Feature_Name, names_to = "Sample_ID", values_to = "Methylation") %>%
  pivot_wider(names_from = Feature_Name, values_from = Methylation)

# Merge feature data with annotation data by Sample_ID
test_data <- ano_test %>%
  inner_join(df_test_feature_t, by = "Sample_ID")

# Extract Gestational Age (GA) as target variable
test_GA <- test_data$Gestational_Age

# Select feature columns (methylation features) for testing
test_features <- test_data %>% select(all_of(feature_names))

# --- Feature Scaling ---
# Scale features using training set statistics (important to avoid data leakage)
feature_scaling_params <- scale(train_features)
train_features_scaled <- as.matrix(feature_scaling_params)
test_features_scaled <- scale(as.matrix(test_features),
                              center = attr(feature_scaling_params, "scaled:center"),
                              scale = attr(feature_scaling_params, "scaled:scale"))

# Convert target variables to numeric vectors
train_GA <- as.numeric(train_GA)
test_GA <- as.numeric(test_GA)


# ------------------- Model Training (Ridge Regression) -------------------

# Use cross-validation to find the optimal lambda for Ridge Regression
set.seed(123) # for reproducibility
cv_model <- cv.glmnet(train_features_scaled, train_GA, alpha = 0, family = "gaussian") # alpha=0 for Ridge
best_lambda <- cv_model$lambda.min

# Train Ridge Regression model with the optimal lambda
ridge_model <- glmnet(train_features_scaled, train_GA, alpha = 0, lambda = best_lambda, family = "gaussian")

# ------------------- Model Prediction and Evaluation -------------------

# Predict on the test set
predictions <- predict(ridge_model, s = best_lambda, newx = test_features_scaled)

# Calculate RMSE
rmse_value <- rmse(test_GA, predictions)
cat("RMSE on Test Set:", rmse_value, "\n")

# ------------------- Scatter Plot -------------------

# Create a data frame for plotting
plot_data <- data.frame(
  Actual_GA = test_GA,
  Predicted_GA = as.vector(predictions) # Convert matrix to vector
)

# Generate scatter plot
scatter_plot <- ggplot(plot_data, aes(x = Actual_GA, y = Predicted_GA)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") + # Add diagonal line
  labs(
    title = "Predicted vs Actual Gestational Age (Test Set)",
    x = "Actual Gestational Age",
    y = "Predicted Gestational Age",
    caption = paste("RMSE =", round(rmse_value, 3)) # Include RMSE in caption/legend
  ) +
  theme_minimal()

# Print the scatter plot
print(scatter_plot)

File: LLMgeneratedCode/gemini-3.py
import pandas as pd
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import numpy as np
from sklearn.preprocessing import StandardScaler

# --- Load Data ---
# Feature data (methylation) - transposed later
df_training_features = pd.read_csv("df_training.csv")
df_test_features = pd.read_csv("df_test.csv")

# Metadata (including gestational age)
ano_training = pd.read_csv("ano_training.csv")
ano_test = pd.read_csv("ano_test.csv")

# --- Data Preprocessing and Merging ---
# Transpose feature data so samples are rows and features are columns
df_training_features_T = df_training_features.set_index('Unnamed: 0').T.reset_index()
df_test_features_T = df_test_features.set_index('Unnamed: 0').T.reset_index()
df_training_features_T.rename(columns={'index': 'Sample_ID'}, inplace=True)
df_test_features_T.rename(columns={'index': 'Sample_ID'}, inplace=True)

# Merge feature data with metadata using Sample_ID
train_df = pd.merge(ano_training, df_training_features_T, on='Sample_ID')
test_df = pd.merge(ano_test, df_test_features_T, on='Sample_ID')

# --- Prepare Data for Modeling ---
# Target variable: gestational_age
y_train = train_df['gestational_age']
y_test = test_df['gestational_age']

# Feature data: Methylation features (all columns except metadata and Sample_ID)
feature_cols = df_training_features_T.columns.drop('Sample_ID') # Get feature names from transposed data
X_train = train_df[feature_cols]
X_test = test_df[feature_cols]

# --- Feature Scaling ---
# Scale methylation features using StandardScaler (important for Ridge)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# --- Model Training ---
# Using Ridge Regression (L2 regularization) - suitable for high-dimensional data
ridge_model = Ridge()
ridge_model.fit(X_train_scaled, y_train)

# --- Prediction on Test Set ---
y_pred = ridge_model.predict(X_test_scaled)

# --- Evaluation ---
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Root Mean Squared Error (RMSE) on Test Set: {rmse:.4f}")

# --- Scatter Plot of Predicted vs Actual GA ---
plt.figure(figsize=(8, 6))
plt.scatter(y_test, y_pred, alpha=0.6)
plt.xlabel("Actual Gestational Age")
plt.ylabel("Predicted Gestational Age")
plt.title("Predicted vs Actual Gestational Age (Test Set)")
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], color='red', linestyle='--') # Diagonal line for perfect prediction
plt.legend([f"RMSE = {rmse:.4f}"], loc='upper left') # Include RMSE in legend
plt.grid(True)
plt.tight_layout()
plt.savefig("./out2/gemini-3.py.pdf", format="pdf")
